Adding initial, prototype implementation for a value type BufferPool

sokket opened the pull request:
Adding initial, prototype implementation for a value type BufferPool.

/cc @KrzysztofCwalina

VSadov commented on src/System.Buffers/src/System/Buffers/BufferData.cs:
```diff
@@ -0,0 +1,20 @@
+    public struct BufferData
+    {
+        internal byte[] _data;
+        internal bool _allocd;
```
I do not think you need to track _allocd. You are using this to not accept back to the pool buffers that did not come from the pool, but I think as long as there is a space in the pool and the buffer is not larger than _maxBufferSize you could just take it. They are all just byte arrays.

sokket replied:
That's a good idea, but what about when all the buffers are returned? Then we would have extra buffers with no space to put them...re-allocing the array for more space would be bad and adding special code to handle that case inside the spin lock seems like it would hurt perf.

Thoughts?

VSadov replied:
I'd just drop extra buffers on the floor. 
You are already dropping buffers when pool is oversubscribed. That is unavoidable since you have a fixed size pool but potentially unbounded demand for buffers (however unlikely).
I only suggest to drop under different conditions - not when a buffer was not from the pool originally, but when pool does not have space to accept it back.

sokket replied:
Cool, good idea...changed to work that way instead

VSadov commented on src/System.Buffers/src/System/Buffers/BufferBucket.cs:
```diff
@@ -0,0 +1,76 @@
+                _lock.Enter(ref taken);
+                Debug.Assert(taken);
+
+                bd._data = _data[_index];
```
Looks like a race. 
_index may have changed since we did  ``` if (_index + 1>= _data.Length)``` above

sokket replied:
Good catch, thanks...fixing now

VSadov commented on src/System.Buffers/src/System/Buffers/BufferBucket.cs:
```diff
@@ -0,0 +1,76 @@
+            _data = new byte[numberOfBuffers][];
+            for (int i = 0; i < numberOfBuffers; i++)
+            {
+                _data[i] = new byte[bufferSize];
```
Seems a bit generous to allocate all buffers upfront. We may need just one or two.
Perhaps these could be lazy allocated in Rent when actually needed?

sokket replied:
I went back and forth on this one and went with the upfront allocation so that all calls to Rent would (assuming we had free buffers of the requested size) be constant-time calls; if we had the buffers be lazy allocated, you would lose efficiency since you would have to go and allocate the buffers on first response. My thought was that it would be better to push the cost up-front and have constant-time rent calls, especially since the Pool itself is an instance-object...so they wouldn't create a pool with certain buffer sizes unless the caller knew they needed one.

bartonjs commented on src/System.Buffers/src/System/Buffers/BufferBucket.cs:
```diff
@@ -0,0 +1,82 @@
+
+            // Be quick and do an unsynchronized read to see if all
+            // of our buffers have been used
+            if (_index + 1 >= _data.Length)
```
Why `+ 1` and `>=`, subtract one from both sides and get `if (_index > _data.Length)`, though since it's about to be a deref maybe you still want `>=`...

sokket replied:
Good point :)

KrzysztofCwalina commented on src/System.Buffers/src/System/Buffers/IBufferPool.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace System.Buffers
+{
+    public interface IBufferPool<T>
```
What's the reason for this abstraction? And if we need an abstraction, why not abstract class?

davidfowl replied:
Because interfaces are cool

sokket replied:
I originally made the abstraction for the ASP.NET case where they look to need byte[] and char[] buffers, but I think I can refactor this PR a little bit more to make it flexible for both without the abstraction

KrzysztofCwalina commented on src/System.Buffers/src/System/Buffers/ByteBufferPool.cs:
```diff
@@ -0,0 +1,86 @@
+            else
+            {
+                buffer = _buckets[SelectBucketIndex(size)].Rent();
+                if (clearBuffer) Array.Clear(buffer, 0, buffer.Length);
```
I wonder if Return should not have clear either or only. I think clear on return is more useful for code that does not want to leak secrets.

sokket replied:
Good idea, I'll add it there. I think it's useful to have in both Rent and Return cases...for instance, if someone rents a buffer, the contract doesn't guarantee that the buffer is empty so if they are going to use the buffer to receive data then they probably want it empty; however, if they are going to fill it then they don't care if it is empty or not. Having a clear on both Rent and Return looks to give the maximum amount of flexibility with very minimal overhead

KrzysztofCwalina commented on src/System.Buffers/src/System/Buffers/ByteBufferPool.cs:
```diff
@@ -0,0 +1,86 @@
+
+            for (int i = 0; i < _buckets.Length; i++)
+            {
+                _buckets[i] = new BufferBucket(GetMaxSizeForBucket(i), NUMBER_OF_BUFFERS_IN_BUCKET);
```
This is a pretty large allocation. Could we make it lazy?

sokket replied:
I went back and forth with the lazy initialization and I went with this way for a couple reasons:

1) Since the buffer pool is an instance class, if someone creates one then they know that they will need the buffers and it isn't initialized for them behind the scenes
2) In step with above, since we know someone needs one since it's created, I optimized for constant time (with the disclaimer of all buffers being full and needing to alloc one) Rent calls; the caller doesn't need to worry about if this Rent call will kick a Lazy initialization or not since the initialization cost is paid during initialization.
3) This is more a guess than the other two, but my guess on the usage pattern of a Buffer Pool is that it will probably be initialized during the program initialization with other book keeping (such as binding and listening to sockets or hitting the disk for I/O) so if the program needs a buffer pool then it's better (IMHO) to upfront the cost so you can have fast access when you really need perf later on doing important work.

Thoughts?

KrzysztofCwalina commented:
Also,

KrzysztofCwalina commented:
closed by mistake. sorry.

bartonjs commented on src/System.Buffers/src/System/Buffers/BufferBucket.cs:
```diff
@@ -0,0 +1,82 @@
+            // then there was a buffer alloc'd that was returned instead so
+            // we can just drop this buffer
+            if (_index == 0)
+                buffer = null;
```
Why bother assigning the input parameter to null?

sokket replied:
Ah I should make this a ref parameter so that the caller can't Return a buffer and then still have a reference and keep using it after returning

KrzysztofCwalina replied:
I had that as a ref parameter before. It does not work really well. It's annoying and it does not allow returning buffers that are properties, e.g. BufferFormatter.Buffer. In the end I changed this API back to not ref.

sokket commented:
Changed the Buffer Pool to be generic to any value type so that it can work with Byte[] or Char[], meaning it should help with ASP.NET. Also addressing feedback

sokket commented:
Any more feedback @bartonjs, @KrzysztofCwalina, @VSadov?

VSadov commented:
@sokket   Honestly, I am not sure I like the lock. Locks have a habit of not bothering you until there is a contention in a real-world scenario causing dramatic perf drops because of convoying.

I'd prefer a lockless approach as something that scales more predictably. Perhaps similar strategy to https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/SharedCollections/ObjectPool%601.cs

However, lock might be ok for initial implementation.

sokket commented:
@VSadov it should be super lightweight since it isn't a full lock, just a SpinWait. Since the code inside the spin is super tiny, it seemed like the simplest approach for the initial implementation until we get the full requirements of how we want this new data structure to look and act.

KrzysztofCwalina commented:
I think it's good for the initial commit, i.e. good to go with me.

sokket commented:
Cool, thanks! Merging
